ipa-core/src/protocol/ipa_prf/aggregation/bucket.rs

```diff
@@ -119,10 +118,20 @@ where
         let mut futures = Vec::with_capacity(breakdown_count / step);
 
         for (i, tree_index) in (0..breakdown_count).step_by(step).enumerate() {
-            let bit_c = depth_c.narrow(&BucketStep::from(i));
+            let bucket_c = depth_c.narrow(&BucketStep::from(i));
+
+            let index_contribution = row_contribution[tree_index].iter();
 
             if robust || tree_index + span < breakdown_count {
-                futures.push(row_contribution[tree_index].multiply(bit_of_bdkey, bit_c, record_id));
+                futures.push(async move {
+                    let bit_futures = index_contribution
+                        .zip(repeat((bit_of_bdkey, bucket_c.clone())))
+                        .enumerate()
+                        .map(|(i, (a, (b, bucket_c)))| {
+                            a.multiply(b, bucket_c.narrow(&BitOpStep::Bit(i)), record_id)
+                        });
```

Comment:
This got a lot more complex looking. 

Does it no longer work to multiply a multi-bit value with a single bit? That's too bad. Could we migrate from BitDecomposed to a BooleanArray and use `select` instead?

Doesn't need to be in this PR - this PR is too long already =).

Reply:
The issue here is that we're working with a 2-D array. The type `BitDecomposed<AdditiveShare<Boolean, N>>` actually does contain a boolean array -- the vector type in `AdditiveShare<Boolean, N>` is the boolean array type of length N.

We could flatten everything into a boolean array of size N * HV_BITS, but then we'd have to do a bunch of 2-D array index calculations, which I think would make things even more complicated.

I do like the idea of using select, though. I'll see if I can do something with that.

Reply:
After looking at it some, I don't think it makes sense to do in this PR... it's a 100 lines elsewhere for 10 lines here tradeoff. Which may be the right tradeoff, if it improves the modularity of the code, but I don't think it's the right tradeoff if those 100 lines are getting added to this PR.

The concrete options I see are:
* Implementing `BooleanArrayMul` for `BitDecomposed<AdditiveShare<Boolean, N>>`. This in turn requires either adding some more `Foo: SecureMul<C>` bounds in various places, or else adding a `C` type parameter to the `BooleanArrayMul` trait.
* Converting `select` into a trait so we can implement it for arbitrary types.

Reply:
I'm totally supportive of landing this PR as is and working on this in a future PR.

Reply:
My plan is to defer this and saturating addition to a follow-up. I am working on the (rather gruesome) merge conflicts presently, but I'm hoping this can be ready tonight or tomorrow.